source/exe/signal_action.cc

```diff
@@ -0,0 +1,74 @@
+#include "exe/signal_action.h"
+
+#include <signal.h>
+
+constexpr int SignalAction::FATAL_SIGS[];
+
+void SignalAction::SigHandler(int sig, siginfo_t* info, void* context) {
+  void* error_pc = 0;
+
+  const ucontext_t* ucontext = reinterpret_cast<const ucontext_t*>(context);
+  if (ucontext != nullptr) {
+#ifdef REG_RIP
+    // x86_64
+    error_pc = reinterpret_cast<void*>(ucontext->uc_mcontext.gregs[REG_RIP]);
+#elif defined(REG_EIP)
+    // x86 Classic - not tested
+    error_pc = reinterpret_cast<void*>(ucontext->uc_mcontext.gregs[REG_EIP]);
+#elif defined(__arm__)
+    // ARM - not tested
+    error_pc = reinterpret_cast<void*>(ucontext->uc_mcontext.arm_pc);
+#elif defined(__ppc__)
+    // PPC - not tested
+    error_pc = reinterpret_cast<void*>(ucontext->uc_mcontext.regs->nip);
+#else
+#warning "Cannot determine PC location in machine context for your architecture"
+#endif
+  }
+
+  BackwardsTrace tracer(true);
+  tracer.LogFault(strsignal(sig), info->si_addr);
+  if (error_pc != 0) {
+    tracer.CaptureFrom(error_pc);
+  } else {
+    tracer.Capture();
+  }
+  tracer.Log();
+
+  signal(sig, SIG_DFL);
+  raise(sig);
+}
+
+void SignalAction::InstallSigHandlers() {
+  stack_t stack;
+  stack.ss_sp = altstack_.get();
+  stack.ss_size = ALTSTACK_SIZE;
+  stack.ss_flags = 0;
+
+  if (sigaltstack(&stack, nullptr) < 0) {
```

Comment:
Looks like mmap with MAP_ANONYMOUS will map a zero'd page, so doing that will address this as well.